File: src/main/java/com/mongodb/spark/MongoRDD.java
/*
 * Copyright (c) 2008-2015 MongoDB, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.mongodb.spark;

import com.mongodb.client.MongoCursor;
import org.apache.spark.Partition;
import org.apache.spark.SparkContext;
import org.apache.spark.TaskContext;
import org.apache.spark.broadcast.Broadcast;
import org.apache.spark.rdd.RDD;
import org.bson.Document;
import org.bson.conversions.Bson;
import scala.collection.Iterator;
import scala.collection.mutable.ArrayBuffer;
import scala.reflect.ClassTag$;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;

import static com.mongodb.assertions.Assertions.isTrueArgument;
import static com.mongodb.assertions.Assertions.notNull;
import static scala.collection.JavaConverters.asScalaIteratorConverter;

/**
 * An RDD that executes queries on a Mongo connection and reads results.
 *
 * @param <T> the type of the objects in the RDD
 */
public class MongoRDD<T> extends RDD<T> {
    private Class<T> clazz;
    private Broadcast<MongoCollectionProvider<T>> collectionProvider;
    private int maxChunkSize;
    private List<Bson> pipeline;
    private String splitKey;

    /**
     * Constructs a new instance. The number of partitions is determined by the number of chunks in a
     * sharded collection, or the number of chunks calculated by a vectorSplit for non-sharded
     * collections (default max chunk size 64 MB).
     *
     * Note: the collection provider will be broadcasted through the input SparkContext.
     *
     * @param sc the spark context the RDD belongs to
     * @param provider the mongo collection provider for the RDD
     * @param clazz the class of the elements in the RDD
     * @param splitKey the minimal prefix key of the index to be used for splitting
     */
    public MongoRDD(final SparkContext sc, final MongoCollectionProvider<T> provider, final Class<T> clazz,
                    final String splitKey) {
        this(sc, provider, clazz, splitKey, 64, Collections.emptyList());
    }

    /**
     * Constructs a new instance. Set maxChunkSize to determine the size of the partitions
     * based on a vectorSplit. maxChunkSize only affects creating RDDs from non-sharded
     * collections.
     *
     * Note: the collection provider will be broadcasted through the input SparkContext.
     *
     * @param sc the spark context the RDD belongs to
     * @param provider the mongo collection provider for the RDD
     * @param clazz the class of the elements in the RDD
     * @param splitKey the minimal prefix key of the index to be used for splitting
     * @param maxChunkSize the max chunk size for partitions in MB
     */
    public MongoRDD(final SparkContext sc, final MongoCollectionProvider<T> provider, final Class<T> clazz,
                    final String splitKey, final int maxChunkSize) {
        this(sc, provider, clazz, splitKey, maxChunkSize, Collections.emptyList());
    }

    /**
     * Constructs a new instance. The number of partitions is determined by the number of chunks in a
     * sharded collection, or the number of chunks calculated by a vectorSplit for non-sharded
     * collections (default max chunk size 64 MB).
     *
     * Note: the collection provider will be broadcasted through the input SparkContext.
     *
     * @param sc the spark context the RDD belongs to
     * @param provider the mongo collection provider for the RDD
     * @param clazz the class of the elements in the RDD
     * @param splitKey the minimal prefix key of the index to be used for splitting
     * @param pipeline the aggregation pipeline
     */
    public MongoRDD(final SparkContext sc, final MongoCollectionProvider<T> provider, final Class<T> clazz,
                    final String splitKey, final List<Bson> pipeline) {
        this(sc, provider, clazz, splitKey, 64, pipeline);
    }

    /**
     * Constructs a new instance. Set maxChunkSize to determine the size of the partitions
     * based on a vectorSplit. maxChunkSize only affects creating RDDs from non-sharded
     * collections.
     *
     * Note: the collection provider will be broadcasted through the input SparkContext.
     *
     * @param sc the spark context the RDD belongs to
     * @param provider the mongo collection provider for the RDD
     * @param clazz the class of the elements in the RDD
     * @param splitKey the minimal prefix key of the index to be used for splitting
     * @param maxChunkSize the max chunk size for partitions in MB
     * @param pipeline the aggregation pipeline
     */
    public MongoRDD(final SparkContext sc, final MongoCollectionProvider<T> provider, final Class<T> clazz,
                    final String splitKey, final int maxChunkSize, final List<Bson> pipeline) {
        super(sc, new ArrayBuffer<>(), ClassTag$.MODULE$.apply(notNull("clazz", clazz)));
        this.clazz = clazz;
        notNull("provider", provider);
        this.collectionProvider = this.sparkContext().broadcast(provider, ClassTag$.MODULE$.apply(provider.getClass()));
        this.splitKey = notNull("splitKey", splitKey);
        isTrueArgument("maxChunkSize in range [1, 1024]", maxChunkSize >= 1 && maxChunkSize <= 1024);
        this.maxChunkSize = maxChunkSize;
        this.pipeline = pipeline == null ? Collections.emptyList() : pipeline;
    }

    @Override
    public Iterator<T> compute(final Partition split, final TaskContext context) {
        MongoCursor<T> cursor = this.getCursor((MongoPartition) split);
        context.addTaskCompletionListener(new MongoTaskCompletionListener(cursor));
        return asScalaIteratorConverter(cursor).asScala();
    }

    /**
     * Helper function to retrieve the results from the collection.
     *
     * @return the cursor
     */
    private MongoCursor<T> getCursor(final MongoPartition partition) {
        Document partitionBounds = partition.getBounds();

        List<Bson> partitionPipeline = new ArrayList<>();
        partitionPipeline.add(new Document("$match", partitionBounds));
        partitionPipeline.addAll(this.pipeline);

        return this.collectionProvider.value().getCollection().aggregate(partitionPipeline, this.clazz).iterator();
    }

    @Override
    public Partition[] getPartitions() {
        final AtomicInteger count = new AtomicInteger(0);

        return this.getSplitBounds().stream()
                                    .map((document) -> new MongoPartition(count.getAndIncrement(), document))
                                    .toArray(Partition[]::new);
    }

    /**
     * Gets the split keys for the mongo collection to optimize RDD calculation.
     *
     * @return the split keys
     */
    private List<Document> getSplitBounds() {
        Document collStatsCommand = new Document("collStats", this.collectionProvider.value()
                                                                                     .getCollection()
                                                                                     .getNamespace()
                                                                                     .getCollectionName());
        Document result = this.collectionProvider.value()
                                                 .getDatabase()
                                                 .runCommand(collStatsCommand, Document.class);

        if (result.get("ok").equals(1.0)) {
            if (Boolean.TRUE.equals(result.get("sharded"))) {
                return new ShardedMongoSplitter(this.collectionProvider.value(), this.splitKey).getSplitBounds();
            } else {
                return new StandaloneMongoSplitter(this.collectionProvider.value(), this.splitKey, this.maxChunkSize).getSplitBounds();
            }
        }
        else {
            throw new SplitException("Could not get collection statistics. Server errmsg: " + result.get("errmsg"));
        }
    }
}

File: src/main/scala/com/mongodb/spark/rdd/MongoRDD.scala
/*
 * Copyright 2016 MongoDB, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.mongodb.spark.rdd

import scala.collection.JavaConverters._
import scala.reflect.ClassTag
import scala.reflect.runtime.universe._
import scala.util.Try
import org.apache.spark.{Partition, SparkContext, TaskContext}
import org.apache.spark.broadcast.Broadcast
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.types.StructType
import org.apache.spark.sql.{DataFrame, Dataset, SparkSession}
import org.bson.conversions.Bson
import org.bson.{BsonDocument, Document}
import com.mongodb.{MongoClient, MongoCursorNotFoundException}
import com.mongodb.client.{AggregateIterable, MongoCursor}
import com.mongodb.spark.config.ReadConfig
import com.mongodb.spark.exceptions.MongoSparkCursorNotFoundException
import com.mongodb.spark.rdd.api.java.JavaMongoRDD
import com.mongodb.spark.rdd.partitioner.{MongoPartition, MongoSinglePartitioner}
import com.mongodb.spark.{MongoConnector, MongoSpark, NotNothing, classTagToClassOf}

import scala.collection.Iterator

/**
 * MongoRDD Class
 *
 * @param connector the [[com.mongodb.spark.MongoConnector]]
 * @param readConfig the [[com.mongodb.spark.config.ReadConfig]]
 * @tparam D the type of the collection documents
 */
class MongoRDD[D: ClassTag](
    @transient val sparkSession:   SparkSession,
    private[spark] val connector:  Broadcast[MongoConnector],
    private[spark] val readConfig: ReadConfig
) extends RDD[D](sparkSession.sparkContext, Nil) {

  @transient val sc: SparkContext = sparkSession.sparkContext
  private def mongoSpark = {
    checkSparkContext()
    MongoSpark(sparkSession, connector.value, readConfig)
  }

  override def toJavaRDD(): JavaMongoRDD[D] = JavaMongoRDD(this)

  override def getPreferredLocations(split: Partition): Seq[String] = split.asInstanceOf[MongoPartition].locations

  /**
   * Creates a `DataFrame` based on the schema derived from the optional type.
   *
   * '''Note:''' Prefer [[toDS[T<:Product]()*]] as computations will be more efficient.
   *  The rdd must contain an `_id` for MongoDB versions < 3.2.
   *
   * @tparam T The optional type of the data from MongoDB, if not provided the schema will be inferred from the collection
   * @return a DataFrame
   */
  def toDF[T <: Product: TypeTag](): DataFrame = mongoSpark.toDF[T]()

  /**
   * Creates a `DataFrame` based on the schema derived from the bean class.
   *
   * '''Note:''' Prefer [[toDS[T](beanClass:Class[T])*]] as computations will be more efficient.
   *
   * @param beanClass encapsulating the data from MongoDB
   * @tparam T The bean class type to shape the data from MongoDB into
   * @return a DataFrame
   */
  def toDF[T](beanClass: Class[T]): DataFrame = mongoSpark.toDF(beanClass)

  /**
   * Creates a `DataFrame` based on the provided schema.
   *
   * @param schema the schema representing the DataFrame.
   * @return a DataFrame.
   */
  def toDF(schema: StructType): DataFrame = mongoSpark.toDF(schema)

  /**
   * Creates a `Dataset` from the collection strongly typed to the provided case class.
   *
   * @tparam T The type of the data from MongoDB
   * @return
   */
  def toDS[T <: Product: TypeTag: NotNothing](): Dataset[T] = mongoSpark.toDS[T]()

  /**
   * Creates a `Dataset` from the RDD strongly typed to the provided java bean.
   *
   * @tparam T The type of the data from MongoDB
   * @return
   */
  def toDS[T](beanClass: Class[T]): Dataset[T] = mongoSpark.toDS[T](beanClass)

  /**
   * Returns a copy with the specified aggregation pipeline
   *
   * @param pipeline the aggregation pipeline to use
   * @return the updated MongoRDD
   */
  def withPipeline[B <: Bson](pipeline: Seq[B]): MongoRDD[D] = copy(readConfig = readConfig.withPipeline(pipeline))

  /**
   * Allows to copying of this RDD with changing some of the properties
   */
  def copy(
    connector:  Broadcast[MongoConnector] = connector,
    readConfig: ReadConfig                = readConfig
  ): MongoRDD[D] = {
    checkSparkContext()
    new MongoRDD[D](
      sparkSession = sparkSession,
      connector = connector,
      readConfig = readConfig
    )
  }

  override protected def getPartitions: Array[Partition] = {
    checkSparkContext()
    try {
      val partitions = readConfig.partitioner.partitions(connector.value, readConfig, readConfig.pipeline.toArray)
      logDebug(s"Created partitions: ${partitions.toList}")
      partitions.asInstanceOf[Array[Partition]]
    } catch {
      case t: Throwable =>
        logError(
          s"""
             |-----------------------------
             |WARNING: Partitioning failed.
             |-----------------------------
             |
            |Partitioning using the '${readConfig.partitioner.getClass.getSimpleName}' failed.
             |
            |Please check the stacktrace to determine the cause of the failure or check the Partitioner API documentation.
             |Note: Not all partitioners are suitable for all toplogies and not all partitioners support views.%n
             |
            |-----------------------------
             |""".stripMargin
        )
        throw t
    }

  }

  override def compute(split: Partition, context: TaskContext): Iterator[D] = {
    val client = connector.value.acquireClient()
    val cursor = getCursor(client, split.asInstanceOf[MongoPartition])
    context.addTaskCompletionListener((ctx: TaskContext) => {
      logDebug("Task completed closing the MongoDB cursor")
      Try(cursor.close())
      connector.value.releaseClient(client)
    })
    MongoCursorIterator(cursor)
  }

  /**
   * Retrieves the partition's data from the collection based on the bounds of the partition.
   *
   * @return the cursor
   */
  private def getCursor(client: MongoClient, partition: MongoPartition)(implicit ct: ClassTag[D]): MongoCursor[D] = {
    val partitionPipeline: Seq[BsonDocument] = if (partition.queryBounds.isEmpty) {
      readConfig.pipeline
    } else {
      new BsonDocument("$match", partition.queryBounds) +: readConfig.pipeline
    }

    logDebug(s"Creating cursor for partition #${partition.index}. pipeline = ${partitionPipeline.map(_.toJson).mkString("[", ", ", "]")}")
    val aggregateIterable: AggregateIterable[D] = client.getDatabase(readConfig.databaseName)
      .getCollection[D](readConfig.collectionName, classTagToClassOf(ct))
      .withReadConcern(readConfig.readConcern)
      .withReadPreference(readConfig.readPreference)
      .aggregate(partitionPipeline.asJava)

    readConfig.aggregationConfig.hint.map(aggregateIterable.hint)
    readConfig.aggregationConfig.collation.map(aggregateIterable.collation)
    aggregateIterable.allowDiskUse(readConfig.aggregationConfig.allowDiskUse)
    readConfig.batchSize.map(s => aggregateIterable.batchSize(s))
    aggregateIterable.iterator
  }

  private case class MongoCursorIterator(cursor: MongoCursor[D]) extends Iterator[D] {
    override def hasNext: Boolean = try {
      cursor.hasNext
    } catch {
      case e: MongoCursorNotFoundException => throw new MongoSparkCursorNotFoundException(e)
    }

    override def next(): D = try {
      cursor.next()
    } catch {
      case e: MongoCursorNotFoundException => throw new MongoSparkCursorNotFoundException(e)
    }
  }

  private def checkSparkContext(): Unit = {
    require(
      Option(sc).isDefined,
      """RDD transformation requires a non-null SparkContext.
        |Unfortunately SparkContext in this MongoRDD is null.
        |This can happen after MongoRDD has been deserialized.
        |SparkContext is not Serializable, therefore it deserializes to null.
        |RDD transformations are not allowed inside lambdas used in other RDD transformations.""".stripMargin
    )
  }

  private[spark] lazy val hasSampleAggregateOperator: Boolean = connector.value.hasSampleAggregateOperator(readConfig)

  private[spark] def appendPipeline[B <: Bson](extraPipeline: Seq[B]): MongoRDD[D] = withPipeline(readConfig.pipeline ++ extraPipeline)
}
